pandas/core/arrays/numpy_.py

```diff
@@ -0,0 +1,446 @@
+        return cls(np.concatenate(to_concat))
+
+    # ------------------------------------------------------------------------
+    # Data
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    # ------------------------------------------------------------------------
+    # NumPy Array Interface
+
+    def __array__(self, dtype=None):
+        return np.asarray(self._ndarray, dtype=dtype)
+
+    _HANDLED_TYPES = (np.ndarray, numbers.Number)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # Lightly modified version of
+        # https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/\
+        # numpy.lib.mixins.NDArrayOperatorsMixin.html
+        # The primary modification is not boxing scalar return values
+        # in PandasArray, since pandas' ExtensionArrays are 1-d.
+        out = kwargs.get('out', ())
+        for x in inputs + out:
+            # Only support operations with instances of _HANDLED_TYPES.
+            # Use PandasArray instead of type(self) for isinstance to
+            # allow subclasses that don't override __array_ufunc__ to
+            # handle PandasArray objects.
+            if not isinstance(x, self._HANDLED_TYPES + (PandasArray,)):
+                return NotImplemented
+
+        # Defer to the implementation of the ufunc on unwrapped values.
+        inputs = tuple(x._ndarray if isinstance(x, PandasArray) else x
+                       for x in inputs)
+        if out:
+            kwargs['out'] = tuple(
+                x._ndarray if isinstance(x, PandasArray) else x
+                for x in out)
+        result = getattr(ufunc, method)(*inputs, **kwargs)
+
+        if type(result) is tuple and len(result):
+            # multiple return values
+            if not lib.is_scalar(result[0]):
+                # re-box array-like results
+                return tuple(type(self)(x) for x in result)
+            else:
+                # but not scalar reductions
+                return result
+        elif method == 'at':
+            # no return value
+            return None
+        else:
+            # one return value
+            if not lib.is_scalar(result):
+                # re-box array-like results, but not scalar reductions
+                result = type(self)(result)
+            return result
+
+    # ------------------------------------------------------------------------
+    # Pandas ExtensionArray Interface
+
+    def __getitem__(self, item):
+        if isinstance(item, type(self)):
+            item = item._ndarray
+
+        result = self._ndarray[item]
+        if not lib.is_scalar(result):
+            result = type(self)(result)
+        return result
+
+    def __setitem__(self, key, value):
+        value = extract_array(value, extract_numpy=True)
+
+        if not lib.is_scalar(key) and is_list_like(key):
+            key = np.asarray(key)
+            if not len(key):
+                # early return to avoid casting unnecessarily.
```

Comment:
this seems duplicative of the below, why is it needed?

Reply:
This is to ensure that we don't coerce types unnecessarily. `__setitem__(PandasArray[int], Any[float])` would normally coerce to float.

```python
In [6]: arr = pd.arrays.PandasArray([1, 2])

In [7]: arr[0] = np.nan

In [8]: arr
Out[8]:
<PandasArray>
[nan, 2.0]
Length: 2, dtype: float64
```

But with this check, we preserve the original dtype

```pytohn
In [9]: arr = pd.arrays.PandasArray([1, 2])

In [10]: arr[[]] = np.nan

In [11]: arr
Out[11]:
<PandasArray>
[1, 2]
Length: 2, dtype: int64
```

Reply:
numpy doesn't allow this, so why are we here? this seems like something we would do for a Series, but these are wrapped numpy arrays.

Reply:
Removed.